### ISSUE TITLE
Split Solo5 into multiple packages

### ISSUE THREAD

**Issue by mato** [Solo5/solo5#487]
This is a general issue to work out two things:

1. Split out tools (`solo5-elftool`, `solo5-virtio-mkimage`, `solo5-virtio-run`) and headers (`include/*`) into separate packages.
2. Do this in a way that will pave the way for OS/distribution-friendly packages of Solo5.

The motivation for 1 are the Mirage 4 cross-compilation improvements (PoC by @TheLortex #486, mirage/ocaml-freestanding#90), the motivation for 2 is to be able to distribute binary unikernels (/cc @hannesm).

**Comment by mato** [comment]
Currently we have multiple `solo5-bindings-*` packages, each conflicting with one another in OPAM, which install various combinations of the following, depending on the actual target XXX:

1. Header files:
    1. Public and private Solo5 headers: `include/solo5/*` **AS** `$PREFIX/include/solo5-bindings-XXX/solo5/*`.
    2. Base C compiler headers appropriated from the host C compiler toolchain at configure time: `include/crt/*` **AS** `$PREFIX/include/solo5-bindings-XXX/crt/*`.
2. Tools:
    1. ELF manifest/ABI tool: `solo5-elftool` **AS** `$PREFIX/bin/solo5-elftool`
    2. Various scripts: `solo5-virtio-mkimage.sh`, `solo5-virtio-run.sh` **AS** `$PREFIX/bin/solo5-YYY`.
3. Bindings:
    1. Compiled bindings as an object file for linking: `solo5_XXX.o` **AS** `$PREFIX/lib/solo5-bindings-XXX/solo5_XXX.o`.
    2. Linker script: `solo5_XXX.lds` **AS** `$PREFIX/lib/solo5-bindings-XXX/solo5_XXX.lds`.
    3. Pkg-config information `solo5-bindings-XXX.pc` **AS** `$PREFIX/lib/pkgconfig/solo5-bindings-XXX.pc`.
4. Tender (hvt and spt only):
    1. Compiled tender binary: `solo5-XXX` **AS** `$PREFIX/bin/solo5-XXX`.

#486 gets us part of the way there, but since we're splitting out packages and thus also incompatibly breaking the build system downstream, we should take this opportunity to go as far as we can given what we know about current and future requirements.

What I _think_ we want to get to is something like:

1. A `solo5-headers` package, which installs (1.i) **AS** `$PREFIX/include/solo5/*` and (1.ii) **AS** `$PREFIX/include/solo5/crt/*` (note the latter is now one level deeper).
2. A `solo5-tools` package, which installs (2.i) and (2.ii) **AS** `$PREFIX/bin/*` (i.e. all "tools" are now in one package).
3. Co-installable `solo5-bindings-XXX` packages, which install (3.i) and (3.ii) **AS** `$PREFIX/lib/solo5/*` (note the directory change) and (3.iii) and (4.i) **AS** before.

In all of the above cases, `$PREFIX` is currently the OPAM prefix, i.e. `opam config var prefix`.

Now, in the OS/distribution package case, imagine that it will be just `/usr/local`, or even `/usr` (eventually in e.g. upstream Debian).

Questions, mainly in the OS/distro packaging context:

1. @hannesm: Do the above paths work for FreeBSD ports?
2. @mato: (Note to self) Do the above paths work for Debian (including upstream, i.e. policy)?
3. The above paths don't allow for co-installation of packages from multiple archs, which may be useful for some cross-compilation/build scenarios. It's possible we don't care, but if we do then arch-dependent parts need to go in arch-specific directories, e.g. `$PREFIX/lib/solo5/x86_64/...`.
4. The above doesn't allow for a co-installation of multiple versions of bindings, tenders or headers. Do we care?
5. Should we also split out the tenders into separate packages? I can see this being useful in the context of shipping binary unikernels, where a user/operator will want a trusted source of e.g. `solo5-hvt` version XXX.
    - If so, what should these be called?
    - If so, shouldn't solo5-bindings be called libsolo5-bindings in an OS/distro (not OPAM) context.
6. Maybe the tenders should go in `/usr/libexec/solo5/...` and a `solo5-common` package depending on them should install some kind of generic `solo5-run` script, but it may be too early to engineer this properly, so just mentioning it for completeness.

**Comment by hannesm** [comment]
Thanks for writing this up. The paths (`$PREFIX/include/solo5` for headers, `$PREFIX/lib/solo5/` for bindings, `$PREFIX/bin` for tools) sound reasonable (pc files from ports on FreeBSD are kept in `$PREFIX/libdata/pkgconfig/`).

"3 co-installation for multiple archs" tbh I'd skip this for now until we need it. In my experience that is either kept under a custom PREFIX (e.g. `$PREFIX/arm64-solo5/` - and then `lib` / `include` etc.) or all binaries have the cross-architecture as part of their name.

"4. multiple versions of bindings, tenders, and headers" I'd as well skip this until the need arises. I remember from earlier discussion that there is version information in manifest, which could be used for selecting the tender - one way to get there would be to install <tender-name>.<manifest-version>, and for convenience also provide a symlink where <tender-name> links to <tender-name>.<latest-manifest-version>.

"5. tenders into separate packages" indeed this is from my point of view important -- the main (only?) user of bindings and headers is opam - and they have to be available on the system you want to compile the unikernel/guest image for. The tender (+eventually the solo5-elftool) is required to execute the unikernel (and independent of opam / pkg-config / ...). So I'm keen on having the tenders as separate package(s). Indeed, I already build FreeBSD solo5 packages (binary package http://198.167.222.206:8080/packages/builder-default/All/solo5-0.6.7.txz) with the attached patch (to only build tools and tenders) and Makefile (I'm not too keen on that patch since it adds more clutter to the existing build system).
[Makefile.txt](https://github.com/Solo5/solo5/files/5658333/Makefile.txt)
[patch-only-tools-and-tenders.txt](https://github.com/Solo5/solo5/files/5658336/patch-only-tools-and-tenders.txt)

I'd be in favour of having different Make targets instead -- `make solo5-elftool` / `make solo5-hvt` / .. / `make test` .. instead of having to plumb different uses through configure options, and `make` compiling and testing everything. Different OS vendors have different packaging practises: separate headers+libraries needed for development from binaries to run it, etc.

From my point of view, the most useful solo5 packages atm would be tenders and elftool - since that's what is needed to execute any unikernel (via a systemd script, an rc script, or albatross).

"6. tenders + solo5-common package" I'm not sure about this.

**Comment by mato** [comment]
@hannesm:

> The paths ($PREFIX/include/solo5 for headers, $PREFIX/lib/solo5/ for bindings, $PREFIX/bin for tools) sound reasonable (pc files from ports on FreeBSD are kept in $PREFIX/libdata/pkgconfig/).

Right, I expected that, so we'll need some extra configurable variables for e.g. `PCFILEDIR`.

> co-installation for multiple archs" tbh I'd skip this for now until we need it. 

Ok, but see below.

> multiple versions of bindings, tenders, and headers" I'd as well skip this 

> I'd be in favour of having different Make targets instead -- make solo5-elftool / make solo5-hvt / .. / make test .. instead of having to plumb different uses through configure options, and make compiling and testing everything. Different OS vendors have different packaging practises: separate headers+libraries needed for development from binaries to run it, etc.

Indeed. The actual split between what gets specified at `configure` vs. selected by `make` targets is a bit unclear to me, depends on the actual implementation.

One thing we should do is standardise on the "usual" use of `PREFIX` and `DESTDIR` in these kinds of packages:

1. `PREFIX` is the path used from the point of view of a "root" somewhere, i.e.in a system-wide install would usually default to `/usr/local` and in an OPAM install to `$(opam config var prefix)`.
2. `DESTDIR` is the path used mainly by (binary) packaging scripts to specify an "staging location" during an _install_ phase. This normally defaults to `""` and is prepended to `PREFIX` for installation only. For example, `PREFIX=/usr/local` `DESTDIR=`/tmp/pkg` results in a hypothetical `make install` installing to `/tmp/pkg/usr/local/...`.

 I'll try a few experiments for both `PREFIX`, `DESTDIR` and a headers / tools / bindings / tenders split of some form and see what I come up with.

**Comment by hannesm** [comment]
> `DESTDIR` / `PREFIX`

I currently have `MAKE_ARGS=	"PREFIX=${STAGEDIR}${PREFIX}"` in the FreeBSD port Makefile to install this into `STAGEDIR`.

**Comment by mato** [comment]
@TheLortex @hannesm:

Okay, I have a work in progress in #488.  TL;DR 90% of the way there **but** it won't actually work.

Differences between #488 and #486:

- `configure.sh` gets an `--enable-targets` option, so `opam` files no longer need to muck with `CONFIG_`. `--enable-targets=none` is what is used for the "common" OPAM package build.
- The Makefiles allow to pick and choose which of { tools, headers, bindings, tenders } you want to build and install, within reason.
- Instead of `solo5-headers`, there is just a "common" `solo5` OPAM package, which installs headers and tools.
- The `solo5-bindings-xxx` OPAM package contains the bindings _and tenders_, if any.
- OS/distribution packagers can (and probably will) make their own choices about package splits (e.g. have a separate `-tenders` package, or have the tenders in the "common" package, or even split headers into a "dev" package, etc.)

This gets us 90% of the way there, **but**:

The way @TheLortex does things in mirage/ocaml-freestanding#90, `FREESTANDING_CFLAGS` is picked up from  `solo5-headers.pc`, which would in this setup be installed by the "common" OPAM package. This is understandable since `ocaml-freestanding` needs to get some common CFLAGS from somewhere, except that in reality it won't work:

1. `genode` uses a different set of CFLAGS including `-fPIC`.
2. ASLR (planned for next year) will require `-fPIE` for the targets that get it (`hvt`, `spt`), but _not_ for the others.

Now what?

1. Either: we mandate that all targets MUST have the same CFLAGS (forever). This implies `genode` will be removed and once ASLR lands all targets will be compiled with `-fPIE`, which **may or may not work**. It also implies that we are stuck if we want to vary CFLAGS for a future target.
2. Or: We figure out some other way to make things work while allowing variations in per-target CFLAGS.

Note that option (2) does _not_ need to be done within the constraints of `pkg-config`; I've wanted to get rid of that for some time, so it could, for example, be replaced with a `solo5-config` script that is more intelligent, generated at configure time and installed as part of the "common" (OPAM) or "dev" (OS) package. However, I still don't see a way to make it work.

**Comment by hannesm** [comment]
Thanks for your work and testing. For me, this read as: CFLAGS (such as -fPIE) are (mirage) target-dependent.

Could the ocaml-freestanding cross-compiler be done without respecting any of these flags - instead put them into the dune workspace context (as it already happens!?), and compile the ocaml runtime (+libm+nolibc) at the time of building the unikernel (when the target is already selected)? I have not looked into the ocaml-freestanding PR to understand whether there are blocking things for such an approach. Also not clear to me what the purpose of the ocam-freestanding cross-compiler would then be.

**Comment by avsm** [comment]
Agreed with @hannesm. Rather than struggling to get a homogenous set of CFLAGS, compiling the runtime with the dune workspace CFLAGS is actually pretty fast, so I'd be in favour of that.

@hannesm then wrote:
> Also not clear to me what the purpose of the ocam-freestanding cross-compiler would then be.
That's the ideal situation :-) As upstream OCaml supports cross-compilation better and better, we should be retiring ocaml-freestanding. This particular approach would get the runtime compiled within a dune workspace, and @dra27 is working steadily on "pure" cross-compilation within the OCaml compiler itself to target native binaries.  Hopefully in a few OCaml releases we won't need freestanding any more.

**Comment by dinosaure** [comment]
> Could the ocaml-freestanding cross-compiler be done without respecting any of these flags - instead put them into the dune workspace context (as it already happens!?), and compile the ocaml runtime (+libm+nolibc) at the time of building the unikernel (when the target is already selected)? I have not looked into the ocaml-freestanding PR to understand whether there are blocking things for such an approach. Also not clear to me what the purpose of the ocam-freestanding cross-compiler would then be.

I'm not sure that it is currently possible to compile the runtime with some specific (and different) flags. I agree that we should not enforce a set of flags for any target (and loose `genode` or use `-fPIE` which may not work).

The current choice about the `ocaml-freestanding` cross-compiler is to compile it and install it only one time (regardless the target choosen by the user) and then use it to compile any MirageOS projects. This choice was made comparing to some others possibilities:
- a 2-step build: building a locally-installed cross-compiler using `duniverse` sources and then building the unikernel
- improve `dune` to perform a single step unikernel build where it takes care of building both the cross-toolchain and the unikernel

These possibilities imply a compilation of the compiler for each MirageOS project - even if it's not really sustainable or scalable, it allows to us to control C flags according to the target choosen by the user.

For my point-of-view, we highlight some limitations and we have these choices:
1) we can consider the work of @dra27 and post-pone/synchronize his work, our ecosystem and ASLR - so nothing change at this time but we will be aware about an other deep change on our stack
2) try to compile the runtime with the `dune` workspace but I let @TheLortex to confirm that it is possible
3) experiment the possibility to compile locally the compiler for each MirageOS project (but I'm not confident with this solution - I'm not sure that we can use `dune-cache` for example to speed-up the compilation)

**Comment by TheLortex** [comment]
Indeed, the cross-compilation story makes sense only if one set of flags is used. As of today, OCaml does not support having multiple targets built-in (like clang) so we have to find something else. But let's keep in mind that in the future, thanks to @dra27's work, we should have a better workflow for cross-compilation. 

I agree with what has been said above : we can't enforce an homogeneous set of flags. I don't think it's a good idea to locally build compilers for each project (quite heavy and not cache-able because the installation prefix changes for each project). So I see two solutions:

1. Evaluate a closed list of flag sets that we need to support all targets (PIC / no-PIC for starters), write these sets of flags as conflicting configuration packages (`ocaml-freestanding-config-PIC`, `ocaml-freestanding-config-no-PIC`), add dependencies to bindings, corresponding to which sets of flags they support (`solo5-bindings-hvt` ->   `...-PIC` | `...-PIC`, `solo5-bindings-genode` -> `...-PIC`). Then ocaml-freestanding would depend on the union of the configuration packages, only one being selected as they all conflict, and would be built with the chosen set of flags. This reduces co-installability, but allows to support a pre-determined set of flags. This requires very small changes in ocaml-freestanding and libc's to fetch the flags from installed configuration packages, but everything else works the same.

2. As suggested by @dinosaure, build libasmrun.a in the dune workspace, meaning that `ocaml-freestanding`, `nolibc` and `openlibm` need to be packaged with dune. I've just checked and it's possible to go this way. The changes would consist in adding a `dune` wrapper that would run the corresponding `Makefile`s and export libraries. With regards to mirage, ocaml-freestanding should be locally fetched by `opam-monorepo` (ex duniverse) and not installed by `opam` anymore. This way the bindings would all be co-installable and in control of the flags.

**Comment by mato** [comment]
> Evaluate a closed list of flag sets that we need to support all targets (PIC / no-PIC for starters), write these sets of flags as conflicting configuration packages (ocaml-freestanding-config-PIC, ocaml-freestanding-config-no-PIC), add dependencies to bindings, [...]

I'm strongly in favour of this option, since it lets us make this work with the existing Mirage 3.x codebase. This means that we can release a split Solo5 and compatible ocaml-freestanding earlier rather than later, and don't have to maintain the older versions in any way. I'm also happy to implement the changes required to get Mirage 3.x to work with this model, since it'd already be much better than what we have today.

Note that `genode` is no longer updated in OPAM since April (see #446). That doesn't change the motivation for being able have different sets of CFLAGS, it just means that at the moment there will be only one. However, for the purposes of this discussion, let's keep using more than one as an example.

Should these "config" OPAM packages be part of `ocaml-freestanding` or part of `solo5`? I'm not sure, but it feels like the "ocaml-freestanding toolchain to select" is a property of the bindings / Solo5 release, and not the other way round.

How about the following, including a whole Mirage 3.x style picture here for clarity:

```
Binding                 Depends on                      Rdepends                Rdepends        Rdepends

solo5-bindings-hvt      solo5-config-ocaml-static
solo5-bindings-spt
solo5-bindings-virtio
solo5-bindings-muen                                     ocaml-freestanding      mirage-solo5    (mirage-generated .opam)
solo5-bindings-xen                                                              mirage-xen

solo5-bindings-genode   solo5-config-ocaml-pic
```

This means that, when e.g. `solo5-bindings-hvt` gets ASLR support, it changes to depend on a new `solo5-config-ocaml-pie`, and so on.

The remaining question in my mind is, what does `solo5-config-*` actually install? A `.pc` file with just some `Cflags` in it? Something else? See my discussion above about replacing pkg-config entirely in this stack with some form of custom `solo5-config` script(s).

**Comment by mato** [comment]
(For the benefit of those reading e-mail notifications: Edited the previous comment to note that `gmp-freestanding` and `zarith-freestanding` also depend on `ocaml-freestanding`, i.e. will get re-built if the "config" gets switched.)

**Comment by mato** [comment]
[Here](https://gist.github.com/mato/c1dd45b01e92859682f2f96078b4297d) is an example of what such a `solo5-config` script could look like. This would be generated by `make` from values determined by `configure.sh` and installed as part of the "common" or "dev" Solo5 OPAM/OS package, whatever that may be.

So you can do things like:
```
$ ./solo5-config.sh --target=hvt --cflags
-ffreestanding -fstack-protector-strong -nostdinc -mstack-protector-guard=global -isystem /usr/local/include/solo5/crt -I/usr/local/include/solo5
```
OR
```
$ ./solo5-config.sh --ocaml=static --cflags
-ffreestanding -fstack-protector-strong -nostdinc -mstack-protector-guard=global -isystem /usr/local/include/solo5/crt -I/usr/local/include/solo5
```

It does mean that the script contains `PREFIX` in it, but that's fine, that just means `--prefix=` becomes a parameter to `configure.sh`, which is not unlike what autotools does.

Thoughts?

**Comment by dinosaure** [comment]
> Thoughts?

As long as you find a realistic solution with @TheLortex, I'm happy with your solution :+1:.

> Downstream, ocaml-freestanding can then EITHER generate its existing .pc file using solo5-config, OR gain a similar ocaml-freestanding-config script and we get to kill pkg-config entirely ❤️

May be we can take this opportunity to delete `pkg-config`, yes.

**Comment by TheLortex** [comment]
I really like the idea of `solo5-config`. Something I'm wondering is how to manage target-specific CFLAGS that are not mandatory to build ocaml-freestanding ? 
I'm thinking of the `xen` target, which installs additional headers that should be made available using an additional `-I` flag. This flag is not necessary to build ocaml-freestanding but useful for all xen-specific dependencies.

**Comment by hannesm** [comment]
I'm wondering about `solo5-config.sh`: isn't this just data (cflags, ldflags) determined completely at (solo5-config) installation time? if this is the case, I don't see a reason to use a shell script (and would use a text file or sexp or rfc822 instead). WDYT? (same comment about `ocaml-freestanding.sh`).

**Comment by mato** [comment]
@TheLortex 

> I'm thinking of the xen target, which installs additional headers that should be made available using an additional -I flag. This flag is not necessary to build ocaml-freestanding but useful for all xen-specific dependencies.

The only other package in the Mirage stack that uses the `xen` headers is `mirage-xen`, and that ships its own identical copy of the Xen headers. Now, I can understand that having two copies of these around in our stack is suboptimal, but if we want to install them, then:

1. They should not be installed **AS** `$PREFIX/include/xen/*` since in an OS package scenario this would conflict with the "normal" Xen headers (e.g. `libxen-dev`).
2. They should not be installed by the Solo5 "common" or "dev" package, as they are only useful to consumers of Solo5/Xen.

...so, that leads me to conclude that the Xen headers should be installed as part of `solo5-bindings-xen`, *AS* `$PREFIX/include/solo5/xen/*`, which means anyone using `solo5-config --cflags` will be able to include these as `<xen/*>` with no extra steps since `-I$PREFIX/include/solo5` will be passed to the compiler.

Right?

@hannesm 

> I'm wondering about solo5-config.sh: isn't this just data (cflags, ldflags) determined completely at (solo5-config) installation time? if this is the case, I don't see a reason to use a shell script (and would use a text file or sexp or rfc822 instead). WDYT? (same comment about ocaml-freestanding.sh).

_Please_, no. I want to be able to call out to it sanely from `Makefile`s or other arbitrary build systems, and not have to parse stuff.

**Comment by dinosaure** [comment]
> > I'm wondering about solo5-config.sh: isn't this just data (cflags, ldflags) determined completely at (solo5-config) installation time? if this is the case, I don't see a reason to use a shell script (and would use a text file or sexp or rfc822 instead). WDYT? (same comment about ocaml-freestanding.sh).
> 
> Please, no. I want to be able to call out to it sanely from Makefiles or other arbitrary build systems, and not have to parse stuff.

And I'm not confident to use RFC822 format.

**Comment by TheLortex** [comment]
> The only other package in the Mirage stack that uses the xen headers is mirage-xen

Yes it is, however it's currently an illusion: all packages that have C stubs (`cstruct`, `io-page`, ...) have them implemented in `mirage-xen` (and `mirage-solo5`). 
I think a proper way to do it would be to have portable C stubs directly in the corresponding packages (`cstruct`, `io-page`) and remove them from `mirage-<target>`. This way `mirage-<target>` base layers wouldn't need to know about C stubs from the whole mirage universe, and it's the job of each package with C stubs to support the mirage targets. Mirage support would be tested within each package repository, thus making them more self-contained. 
In this case we might need to add a `-DMIRAGE_<target>` flag to allow developers to add target-specific C stubs implementations. 

> ...so, that leads me to conclude that the Xen headers should be installed as part of `solo5-bindings-xen`, **AS** `$PREFIX/include/solo5/xen/*`, which means anyone using `solo5-config --cflags` will be able to include these as `<xen/*>` with no extra steps since `-I$PREFIX/include/solo5` will be passed to the compiler.
> 
> Right?

This looks good to me. But I think we need a `-DMIRAGE_XEN` additional flag for downstream packages that would need a xen-specific implementation, as seen above.

What do you think ?

**Comment by mato** [comment]
> Yes it is, however it's currently an illusion: all packages that have C stubs (cstruct, io-page, ...) have them implemented in mirage-xen (and mirage-solo5).

Yes, I know. This is partly a historical accident (back from the old `mirage-platform` days), and partly laziness.

> I think a proper way to do it would be to have portable C stubs directly in the corresponding packages (...)

I've proposed something like this myself previously, but today I'm not so sure. The current arrangement with `mirage-{solo5,xen}` OPAM package containing the stubs is practical, and not a huge maintenance burden. Also, I think we all agree, that we don't particularly want to encourage more packages with C stubs in them.

If we do go down that route it'll be a lot of juggling of dependencies across several more packages. Before we do that, I'd like to see a proper table of what stubs we have in `mirage-{xen,solo5}` today, and where they would live in the new model.

> This looks good to me. But I think we need a -DMIRAGE_XEN additional flag for downstream packages that would need a xen-specific implementation, as seen above.

This is actually orthogonal to "which OPAM package the stubs live in", and is something I'd like to avoid unless we find that we actually need it, on the basis that "No is temporary but yes is forever". Xen is the only Solo5 target which is "special", and I _think_ that the only downstream that needs to know about that is `mirage-xen`, which is self-evident and doesn't require a `-DMIRAGE_XEN` to achieve.

Thoughts? /cc @avsm @samoht

**Comment by dinosaure** [comment]
> Also, I think we all agree, that we don't particularly want to encourage more packages with C stubs in them.

If we want to focus on the IoT perspective (which is not mine), the question behind C stubs becomes more important. I agree that, currently, we have only few packages with C stubs (mostly crypto packages where performance matters). But for IoT, I suspect that the law of rewrite our world only OCaml does not scale up 😕 - and we need to integrate more C code.

**Comment by hannesm** [comment]
@TheLortex 
> Yes it is, however it's currently an illusion: all packages that have C stubs (cstruct, io-page, ...) have them implemented in mirage-xen (and mirage-solo5).

@mato
> Yes, I know. This is partly a historical accident (back from the old mirage-platform days), and partly laziness.

There are counterexamples: digestif, mirage-crypto, hacl, fiat, bigstringaf...

@TheLortex 
> I think a proper way to do it would be to have portable C stubs directly in the corresponding packages (...)

@mato
> I've proposed something like this myself previously, but today I'm not so sure. The current arrangement with mirage-{solo5,xen} OPAM package containing the stubs is practical, and not a huge maintenance burden. Also, I think we all agree, that we don't particularly want to encourage more packages with C stubs in them.

I worked on this topic several years back, and in MirageOS 3.x days this turns out to be complicated (due to the need of some more rules and hacks in the build system).

@mato
> If we do go down that route it'll be a lot of juggling of dependencies across several more packages. Before we do that, I'd like to see a proper table of what stubs we have in mirage-{xen,solo5} today, and where they would live in the new model.

As a release manager of MirageOS 3.x I'd as well prefer to keep the "split solo5 into multiple packages" as the first goal -- which is AFAICS completely MirageOS 3.x compatible. Designing the dependency hierarchy (and who installs the xen headers) should IMHO be done subsequently in a different PR series.

**Comment by mato** [comment]
> As a release manager of MirageOS 3.x I'd as well prefer to keep the "split solo5 into multiple packages" as the first goal -- which is AFAICS completely MirageOS 3.x compatible. Designing the dependency hierarchy (and who installs the xen headers) should IMHO be done subsequently in a different PR series.

Agreed.

To summarize, next steps:

1. Let's defer discussion of "what to do with C stubs that live in `mirage-{solo5,xen}`", as that is orthogonal to the core changes being discussed here.
2. I'll work on updating #488 on the basis of what has been agreed so far:
    - Overall design, "config" packages to deal with `CFLAGS` (https://github.com/Solo5/solo5/issues/487#issuecomment-742537414)
    - Install a `solo5-config` script instead of `pkgconfig`
    - Xen headers get installed by `solo5-bindings-xen` (https://github.com/Solo5/solo5/issues/487#issuecomment-743245284)
3. @TheLortex then updates mirage/ocaml-freestanding#90 to match.
4. I'll put together Mirage 3.x PR(s) to match.

This should get us to a point where we can release Solo5 0.7.0 and ocaml-freestanding 0.7.0, both of which will work with Mirage 3.x and master.

Please speak up if I've missed anything.

**Comment by mato** [comment]
Following up on today's video discussion, I'd like to remove the Genode bindings entirely as part of this work as it's adding unnecessary complexity. We can always re-instate it at a later date once the OCaml cross-compilation situation improves and we have proper support for external toolchains. @ehmry Any objections? Are you (or anyone in the Genode community) actively using the Solo5/MirageOS support?

**Comment by ehmry** [comment]
Removing Genode support is fine by me, it should be easy enough to add
again later.

**Comment by mato** [comment]
@TheLortex Okay, I have #488 95% done, barring some polish and proper commit history.

As per our discussion, there are now new `solo5` and `solo5-config-ocaml-static` OPAM packages, and the bindings depend on both. The pkg-config stuff is gone, and there is now a solo5-config:

```
$ solo5-config --help
usage: solo5-config --target=TARGET | --toolchain=TOOLCHAIN [ OPTIONS ]

Queries Solo5 toolchain parameters.

The Solo5 target to use must be selected using --target=TARGET, where TARGET is
one of 'hvt,spt,virtio,muen,xen'.

Alternatively, if you are configuring an ocaml-freestanding toolchain,
compile-time (but not link-time) parameters may be queried by selecting it with
--toolchain=TOOLCHAIN, where TOOLCHAIN is currently 'static'.

Options:
    --cc
    --ld
    --cflags
    --ldflags
    --lddeps
```

... which should be self-explanatory.

Please take a look, and let me know how you get on.

**Comment by TheLortex** [comment]
This looks good, thank you. So when configuring ocaml-freestanding I should call `solo5-config --toolchain=static` to get the correct flags to build libasmrun.a. This means that I need to find which `solo5-config-ocaml-...` is installed to perform the correct call to `solo5-config`. 
I'm not sure what's the best way to have that:
- probing with `opam list` to find the installed package.
- add a way to list configured toolchains in `solo5-config`.
- have a `solo5-config --toolchain` command which fails if multiple toolchains are possible.

**Comment by mato** [comment]
@TheLortex Good point. Looking through the OPAM manual a bit, I think we could do this with flags to ocaml-freestanding's `configure.sh` and per-package variables. That feels cleaner than having `solo5-config` "know" about co-installability of toolchains, since that's an ocaml-specific thing, at least for the moment.

How about this:

1. Add a required `--toolchain=TOOLCHAIN` flag to ocaml-freestanding's `configure.sh`, where the only currently accepted `TOOLCHAIN` is `static`. (Feel free to swipe the option parsing code from the Solo5 `configure.sh` or `solo5-config.sh`.)
2. Adapt the ocaml-freestanding build to require running `configure.sh` before `make`. (Again, see how Solo5 does this.)
3. In the ocaml-freestanding `opam` file, use something like:
```
build: [ "./configure.sh" "--toolchain=static" { solo5-config-ocaml-static:installed } ]
```
4. While you're at it, you could add a `--prefix=` to ocaml-freestanding's `configure.sh` for consistency, rather than passing it down through `make` variables as we do right now.

**Comment by mato** [comment]
... if I understand the OPAM manual correctly, following the above, when we have multiple toolchains (and `solo5-config-ocaml-X`) packages, we could then do:

```
build: [ "./configure.sh" "--toolchain=static" { solo5-config-ocaml-static:installed } "--toolchain=pie" { solo5-config-ocaml-pie:installed } ]
```

which effectively means the _presence_ of the `solo5-config-ocaml-X` package _selects_ which `--toolchain=` flag is passed to ocaml-freestanding. This looks like an elegant solution.

Side note: Currently `solo5-bindings-*` conflict with the old `solo5-bindings-* { < "0.7.0" }` packages, I think I can simplify this and make just the `solo5-config-ocaml-static` package contain the conflicts. Easier to maintain that way and less duplication.

**Comment by mato** [comment]
Over the last couple of weeks after discussion with @TheLortex I've come to the conclusion that the solution proposed here, and implemented in #488 and mirage/ocaml-freestanding#90, is fundamentally wrong and not thought through. This has become more obvious in the presence of @TheLortex's and my experiments with enabling _cross-architecture_ compilation in both Solo5 and ocaml-freestanding.

How we got here:

1. Solo5, when compiled on a supported _host_ system, builds and installs common headers and _host_ tools_, _some bindings_ and _some or none_ tenders.
2. #488 introduces the concept of a Solo5 _toolchain_, exposed via a `solo5-config` script, and also defines `make install` targets using a well-defined filesystem hierarchy, allowing _some bindings_ to be co-installed in OPAM.
3. #488 splits Solo5 into several OPAM packages: `solo5` (common headers and _host_ tools), `solo5-bindings-XXX` (one for each of _some bindings_, may include a _tender_) and `solo5-config-ocaml-static` (used by ocaml-freestanding to select the _toolchain_).

I believe that the fundamental issue with the above is that the package split in point (3) is wrong, more so in the context of the availability of a Solo5 _cross toolchain_. To explain why, we need to define what a Solo5 _toolchain_ actually consists of:

1. Common headers.
2. Suitable _host_ compiler (not part of Solo5) and flags to compile freestanding code, detected at Solo5 build time.
3. _host_ tools.
4. _some bindings_, compiled for _host_.
5. Metadata to expose all of the above downstream via `solo5-config`, so that consumers can compile C code with (2)  and link against (4).

In the case of a Solo5 _cross toolchain_ for a _target_, this becomes:

6. Common headers.
7. Suitable _cross_ compiler from _host_ to _target_ (not part of Solo5) and flags to compile freestanding code, detected at Solo5 build time.
8. _host_ tools.
9. _some bindings_, cross-compiled for _target_.
10. Metadata to expose all of the above downstream via `solo5-config`, so that consumers can compile C code with (7)  and link against (9).

It follows from the above that (1) and (6) and (3) and (8) are equivalent, i.e. _host_ tools and headers should be built from the same OPAM package only once. It also follows that (2) determines (4) and (5) for a simple _toolchain_, while (7) determines (9) and (10) for a _cross toolchain_.

Fundamentally, the way #488 does things it is trying to express the above in terms of dependency and availability relationships between both the wrong set of, and too many different OPAM packages. With a _cross toolchain_, this would result in an explosion of yet more packages, e.g. `solo5-bindings-XXX-aarch64`.

Therefore, I believe we need a different, simpler, package split for OPAM. The following is a proposal which could work:

1. `solo5`: Installs _host_ tools, common headers, _some bindings_ (as available for _host_), _some or none_ tenders (as available for _host_). Installs metadata describing a `static` _toolchain_. Availability: Union of all current _solo5-bindings-XXX_ constraints.
2. `solo5-config-ocaml-static`: Depends on `solo5 {=version}`, selects `static` toolchain for `ocaml-freestanding`. Conflicts with (4).
3. `solo5-aarch64`: Depends on `solo5 {=version}` (for _host_ tools, common headers), additionally installs _some bindings_ cross-compiled for _target_ (`aarch64`). Installs metadata describing a `static-aarch64` _cross toolchain_. Availability: TBD, e.g. on Debian this would need a depext on `gcc-aarch64-linux-gnu`.
4. `solo5-config-ocaml-static-aarch64`: Depends on `solo5-aarch64 {=version}`, selects `static-aarch64` toolchain for `ocaml-freestanding`. Conflicts with (2).
5. `ocaml-freestanding`: Depends on `solo5_config_ocaml_static | solo5_config_ocaml_static_aarch64`.
6. `mirage-solo5`, `mirage-xen`: Depend on `solo5` and `ocaml-freestanding`.
7. Unikernel: Depends on `mirage-solo5` or `mirage-xen` only, but see below.

Pros (compared to the current setup):

1. Less OPAM packages.
2. `solo5` is installable as a single package, installs all _bindings_, _tenders_ and _toolchains_ available on _host_.
3. `solo5-TARGET` is installable as a single package, installs all _bindngs_ and _cross toolchains_ available on _host_ for _target_.
4. _cross-architecture_ compilation is possible.

Cons (compared to the current setup):

1. The `mirage` workflow would need to change. Currently, `mirage configure -t <any solo5 TARGET>` always works, and the following `make depend` step either "installs Solo5" (via a `solo5-bindings-TARGET` dependency in the generated opam file) or fails if `solo5-bindings-TARGET` is not installable on the _host_. With this proposal, OPAM only knows that "Solo5 is installable", which means that _some bindings_ are available. I can think of the following alternative workflow:
    - `mirage configure -t <any solo5 TARGET> --toolchain TOOLCHAIN` (where _toolchain_ defaults to `static`) causes the `mirage` tool to a) check that `solo5-config-ocaml-TOOLCHAIN` is installed, and if not, prompt the user and install it if confirmed.
    - if the above succeeds, `mirage` then calls out to the now installed `solo5-config` to verify that the combination of `--toolchain=TOOLCHAIN --target=TARGET` is actually available on the _host_. If this fails then a friendly error is presented to the user.
    - If the above succeeds, `mirage` configures the unikernel to build for TARGET and TOOLCHAIN. Given that both the correct `solo5-config-ocaml-TOOLCHAIN` and `solo5-TARGET` package are already installed, `make depend` will proceed to install `ocaml-freestanding` and `mirage-solo5` or `mirage-xen` appropriately.
2. `gmp-freestanding` and `zarith-freestanding` need to be updated to understand/work with a _cross toolchain_; this is the case for any proposal that allows for _cross-architecture_ compilation.

**Comment by TheLortex** [comment]
Hi, I agree with you that something is wrong in the current proposal. 
However I don't think we should drop the `solo5-bindings-XXX` packages. 
They are still useful as they allow to describe the availability of the bindings.

I propose the following package split:

- `solo5-config-ocaml-static` | `solo5-config-ocaml-pie`: conf packages describing conflicting sets of flags.
- `solo5-config-host` | `solo5-config-cross-aarch64` conf packages describing eventual cross-arch compilation.
- `solo5-config`: depends on (`solo5-config-host` | `solo5-config-cross-aarch64`) & (`solo5-config-ocaml-static` | `solo5-config-ocaml-pie`) 
  configures solo5 with the selected options. Only installs `solo5-config`, which should contain all the information needed to build solo5 tools, bindings and tenders. 
- `solo5`: depends on `solo5-config`: installs host tools and common headers.
- `solo5-bindings-XXX`: depends on `solo5-config`: installs bindings and tenders. 
- `ocaml-freestanding`: depends on `solo5` (and `solo5-config`). 

Configuration is done by selecting `solo5-config-*` packages, which can be done manually or generated by the mirage tool. Moreover it should be easy to express availability:
- `solo5-config-cross-aarch64`: depends on `conf-gcc-linux-gnu-aarch64`.
- `solo5-config-host`: available on `(arch = "x86_64" | arch = "arm64") & (os = "linux" | os = "freebsd" | os = "openbsd")`
- `solo5-bindings-(muen|xen|virtio)`: depends on `solo5-config-host`. available on `arch = "x86_64"`.
- `solo5-bindings-genode`: depends on `solo5-config-ocaml-pie`.

In comparison to @mato's proposal, this allows to select a solo5 target and let the opam solver find a configuration that builds that target.

**Comment by mato** [comment]
Together with @TheLortex, after more discussion and iteration on both our proposals we have a third proposal of the Solo5 and ocaml-freestanding OPAM packaging split that we have decided to go for.

Basically we have three cursors to move:
- quantity of opam packages
- compilation time
- co-installability

This proposal is  a trade-off between quantity of packages and compilation time, while maximizing co-installability.

### Goals

- support multiple combinations of Solo5 toolchains (host or cross-ARCH)
- each Solo5 toolchain provides _some_ set of bindings
- least amount of maintenance going forward, with flexibility to add/remove Solo5 toolchains and bindings
- non-goal: no support for cross-arch compilation with Mirage 3

### OPAM package split

`solo5`: Installs all of Solo5 needed to build and run unikernels on the _host_:
- common:
        - headers
        - tools for _host_
        - tenders for _host_
- toolchain(s) for _host_:
        - _some_ bindings for _host_ architecture
        - i.e. currently a single toolchain called `static`

`solo5-cross-ARCH`: Installs a Solo5 _cross_ toolchain targetting ARCH:
- _cross_ toolchain(s) for ARCH:
        - _some_ bindings cross-built for ARCH
        - i.e., using `aarch64` as an example, currently a single toolchain called `static-cross-aarch64`.
- depends on `{solo5 =version}` (for "common" components)
- depext or dependency on `conf-gcc-cross-ARCH`, or some similar way to pull in the appropriate C cross compiler.

`ocaml-freestanding`: Installs all OCaml toolchains for installed Solo5 _host_ toolchains:
- builds and installs a separate freestanding sysroot for each Solo5 _host_ toolchain
- depends on `solo5`

`ocaml-freestanding-cross-ARCH`: Installs all OCaml toolchains for installed Solo5 _cross_ toolchains targetting ARCH:
- builds and installs a separate freestanding sysroot for each Solo5 _cross_ toolchain for ARCH
- depends on `solo5-cross-ARCH`

### Mirage tool workflows

Mirage 4:
- at configure time:
        - User: `mirage configure -t hvt [ --cross=ARCH ]`
        - tool queries via `solo5-config`, asking "is hvt available [ for cross-compilation to ARCH ]"?
- possible outcomes:
        - no solo5-config -> prompt user to install Solo5
        - not available -> error
        - available -> returns Solo5 toolchain / `ocaml-freestanding` sysroot to use
- no more use of `pkg-config` at all (updates to `gmp-freestanding`, `zarith-freestanding` for Mirage 4 to make this work have already been done by @TheLortex)

Mirage 3:
- as for Mirage 4, but no support for multiple toolchains or _cross_ toolchains.
- i.e. fail if the toolchain returned by `solo5-config` for any Solo5 target is not `static`
- still uses `pkg-config` to get the single toolchain's flags from `ocaml-freestanding` (`gmp-freestanding`, `zarith-freestanding` do not change)
- generated unikernel opam file depends on:
        - `ocaml-freestanding` (i.e. host toolchain support only)
        - `mirage-solo5` OR `mirage-xen`.

### Advantages

- Solo5 (`solo5-config`) is the single "source of truth" for "which toolchains and bindings are available"
- No need to maintain the answer to the above as a complex web of OPAM packages and interdependencies
- Least amount of OPAM packages
- All toolchain and bindings combinations are co-installable (though installing "everything" will substantially increase build time of `ocaml-freestanding`).
- Mirage user gets immediate feedback at configure time on validity of target / cross-arch options.

**Comment by mato** [comment]
After much more work together with @TheLortex. we finally have something that works. OPAM-wise this is similar to the last proposal here (https://github.com/Solo5/solo5/issues/487#issuecomment-780611733), but instead of using a `solo5-config` script it introduces new Solo5 _toolchain_ wrappers.

See #494 for details. I'm closing #488 and #486 as this supersedes both.

**Comment by dinosaure** [comment]
#494 was merged via #506. I'm closing this issue.
